libs/logstream/streamer.go

```diff
@@ -0,0 +1,539 @@
+			return err
+		}
+	}
+	b.lines = slices.Clip(b.lines[:0])
+	return nil
+}
+
+func formatTimestamp(ts float64) string {
+	if ts <= 0 {
+		return "----------"
+	}
+	sec := int64(ts)
+	nsec := int64((ts - float64(sec)) * 1e9)
+	t := time.Unix(sec, nsec).UTC()
+	return t.Format(time.RFC3339)
+}
+
+func (s *logStreamer) ensureToken(ctx context.Context) error {
+	if s.token != "" || s.tokenProvider == nil {
+		return nil
+	}
+	token, err := s.tokenProvider(ctx)
+	if err != nil {
+		return err
+	}
+	s.token = token
+	return nil
+}
+
+func (s *logStreamer) refreshToken(ctx context.Context) error {
+	if s.tokenProvider == nil {
+		return errors.New("token refresh unavailable")
+	}
+	s.token = ""
+	return s.ensureToken(ctx)
+}
+
+func decorateDialError(err error, resp *http.Response) error {
+	if resp == nil {
+		return err
+	}
+
+	var bodySnippet string
+	if resp.Body != nil {
+		data, readErr := io.ReadAll(io.LimitReader(resp.Body, handshakeErrorBodyLimit))
+		_ = resp.Body.Close()
+		if readErr == nil {
+			bodySnippet = strings.TrimSpace(string(data))
+		}
+	}
+
+	status := strings.TrimSpace(resp.Status)
+	if status == "" && resp.StatusCode != 0 {
+		status = fmt.Sprintf("%d %s", resp.StatusCode, http.StatusText(resp.StatusCode))
+	}
+	if status == "" {
+		status = "unknown status"
+	}
+
+	detail := fmt.Sprintf("HTTP %s", status)
+	if bodySnippet != "" {
+		return fmt.Errorf("%w (%s: %s)", err, detail, bodySnippet)
+	}
+	return fmt.Errorf("%w (%s)", err, detail)
+}
+
+func (s *logStreamer) shouldRefreshForStatus(resp *http.Response) bool {
+	if resp == nil {
+		return false
+	}
+	switch resp.StatusCode {
+	case http.StatusUnauthorized, http.StatusForbidden:
+		return true
+	default:
+		return false
+	}
+}
+
+func (s *logStreamer) shouldRefreshForError(err error) bool {
+	var closeErr *websocket.CloseError
+	if errors.As(err, &closeErr) {
+		switch closeErr.Code {
+		case closeCodeUnauthorized, closeCodeForbidden:
+			return true
+		}
+	}
+	return false
+}
+
+func handleCloseError(err error) (bool, error) {
+	var closeErr *websocket.CloseError
+	if !errors.As(err, &closeErr) {
+		return false, err
+	}
+	if closeErr.Code == websocket.CloseNormalClosure || closeErr.Code == websocket.CloseGoingAway {
+		return true, nil
+	}
+	return true, fmt.Errorf("log stream closed with code %d (%s): %w", closeErr.Code, strings.TrimSpace(closeErr.Text), err)
+}
+
+func waitForBackoff(ctx context.Context, timer *time.Timer, d time.Duration) error {
+	if d <= 0 {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			return nil
+		}
+	}
+	resetTimer(timer, d)
+	select {
+	case <-ctx.Done():
+		stopTimer(timer)
+		return ctx.Err()
+	case <-timer.C:
+		return nil
+	}
+}
+
+func stopTimer(timer *time.Timer) {
+	if timer == nil {
+		return
+	}
+	if !timer.Stop() {
+		drainTimer(timer)
+	}
+}
+
+func resetTimer(timer *time.Timer, d time.Duration) {
+	if timer == nil {
+		return
+	}
+	if !timer.Stop() {
+		drainTimer(timer)
+	}
+	timer.Reset(d)
+}
```

Comment:
those functions look super similar, can't we have just one, `stopTimer`? `resetTimer` is called only once and it could use `stopTimer` with another line `timer.Reset(d)`.